### PR TITLE
Add st.experimental_user with SiS versioning

### PR DESCRIPTION
**Pull request by sfc-gh-dmatthews**
## 📚 Context
This PR is the final piece to publish a SiS version in the docs and includes the feature st.experimental_user which was previously not versioned.

## 🧠 Description of Changes
* st.experimental_user is now versioned
* All versioned function include an option to select Streamlit in Snowflake. This reflects excluded and modified functions.

## 💥 Impact
Not small

## 🌐 References
* Notion task: [Add SiS version to docs](https://www.notion.so/snowflake-corp/Add-SiS-version-to-doc-30330ce0884f4828bab10c83b8f15614?pvs=4)
* Notion task: [Docs for st.experimental_user for SiS](https://www.notion.so/snowflake-corp/Docs-for-st-experimental_user-for-SiS-54133328c46d47c494edf41a1f9cf2ad?pvs=4)

**Contribution License Agreement**
By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

### CODE REVIEW

**Comment by sfc-gh-jcarroll** [comment]
Update to AppTest and experimental_user looks good to me. I spot checked a few other pages in the SiS version and it mostly looked good. A couple things I noticed which I don't think are merge blockers:
* st.cache_data is listed as not existing in SiS. It does exist, it just doesn't persist cache across sessions
* experimental_rerun has a deprecation notice mentioning 1.27, which is not available to SiS users. That might be confusing.

I didn't review the rest in detail (generate.py code changes etc). But I don't mind clicking the approve button if needed, LMK.

File: python/generate.py [hunk 1]
```diff
@@ -456,8 +458,10 @@ def get_obj_docstring_dict(obj, key_prefix, signature_prefix, only_include=None)
                 is_property,
             )
 
-        # Add the extracted metadata to obj_docstring_dict
-        obj_docstring_dict[fullname] = member_docstring_dict
+        # Add the extracted metadata to obj_docstring_dict if the object is
+        # local to streamlit (null source occurs when the object is inherited
+        if member_docstring_dict["source"]:
+            obj_docstring_dict[fullname] = member_docstring_dict
```
Comment (snehankekre) [inline python/generate.py:464]: What is `member_docstring_dict["source"]:` doing? From going through the code, it looks like `member_docstring_dict` does not have a source attribute (it was never assigned; it's assigned to `description`).

Wouldn't running `python generate.py` result in:

```sh
Traceback (most recent call last):
  File "/docs/python/generate.py", line 551, in <module>
    data = get_streamlit_docstring_dict()
  File "/docs/python/generate.py", line 533, in get_streamlit_docstring_dict
    module_docstring_dict.update(get_obj_docstring_dict(obj, *key))
  File "/docs/python/generate.py", line 463, in get_obj_docstring_dict
    if member_docstring_dict["source"]:
KeyError: 'source'
```

Comment (snehankekre) [reply python/generate.py:464]: Ah, I see. The script doesn't error out only when run on the https://github.com/streamlit/streamlit/tree/docs/docstring-edits branch.

**Review by snehankekre — COMMENTED** [review]
Just noticed that when I run `python generate.py 1.30.0` when I'm using streamlit from the `docs/docstring-edits` branch from core, that `st.experimental_user` contains some additional methods. However, they don't appear in the deploy preview. 

Question: for the deploy preview, did you have to manually remove these methods from streamlit.json? Or am I missing something?

image

File: python/generate.py [hunk 2]
```diff
@@ -456,8 +458,10 @@ def get_obj_docstring_dict(obj, key_prefix, signature_prefix, only_include=None)
                 is_property,
             )
 
-        # Add the extracted metadata to obj_docstring_dict
-        obj_docstring_dict[fullname] = member_docstring_dict
+        # Add the extracted metadata to obj_docstring_dict if the object is
+        # local to streamlit (null source occurs when the object is inherited
+        if member_docstring_dict["source"]:
```
[the remark this answers is not in the document]
Comment (sfc-gh-dmatthews) [reply python/generate.py:463]: An empty source is a failsafe when an inherited method exists. This effectively throws out inherited methods as commented.

File: python/generate.py [hunk 3]
```diff
@@ -146,6 +146,8 @@ def get_docstring_dict(
                 f"{signature_prefix}.{method_name}",
                 is_class_method=True,
             )
+            if meth_obj["source"] == "":
+                continue
```
[the remark this answers is not in the document]
Comment (sfc-gh-dmatthews) [reply python/generate.py:150]: Necessary to throw out inherited methods. This is a result of documenting more things like AppTesting, query_params, and experimental_user. There is likely a more robust solution but failure to identify the source seems to work thus far in identifying those inherited methods.

**Comment by sfc-gh-dmatthews** [comment]
> Question: for the deploy preview, did you have to manually remove these methods from streamlit.json? Or am I missing something?

The methods are removed automatically from the checks for empty source mentioned above. If they are showing up for you, I wonder if there's a difference in dependencies. I can try to rebuild everything cleanly to check.

**Comment by snehankekre** [comment]
> If they are showing up for you, I wonder if there's a difference in dependencies. I can try to rebuild everything cleanly to check.

@sfc-gh-dmatthews I just started from a clean env (using the `docs/docstring-edits` branch) and ran `python generate.py 1.30.0` and ran into the same issue again.

<details><summary>Here's the relevant part of the overwritten json file</summary>

```json
 "streamlit.experimental_user.get": {
            "name": "experimental_user.get",
            "signature": "st.experimental_user.get(key, default=None)",
            "description": "<p>D.get(k[,d]) -&gt; D[k] if k in D, else d.  d defaults to None.</p>\n",
            "args": [],
            "returns": [],
            "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/../../../../../../../Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/_collections_abc.py#L760"
        },
        "streamlit.experimental_user.items": {
            "name": "experimental_user.items",
            "signature": "st.experimental_user.items()",
            "description": "<p>D.items() -&gt; a set-like object providing a view on D's items</p>\n",
            "args": [],
            "returns": [],
            "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/../../../../../../../Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/_collections_abc.py#L779"
        },
        "streamlit.experimental_user.keys": {
            "name": "experimental_user.keys",
            "signature": "st.experimental_user.keys()",
            "description": "<p>D.keys() -&gt; a set-like object providing a view on D's keys</p>\n",
            "args": [],
            "returns": [],
            "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/../../../../../../../Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/_collections_abc.py#L775"
        },
        "streamlit.experimental_user.to_dict": {
            "name": "experimental_user.to_dict",
            "signature": "st.experimental_user.to_dict()",
            "description": "<p>Get user info as a dictionary.</p>\n<p>This method primarily exists for internal use and is not needed for\nmost cases. <tt class=\"docutils literal\">st.experimental_user</tt> returns an object that inherits from\n<tt class=\"docutils literal\">dict</tt> by default.</p>\n",
            "args": [],
            "returns": [
                {
                    "type_name": "Dict[str,str]",
                    "is_generator": false,
                    "description": "<p>A dictionary of the current user's information.</p>\n",
                    "return_name": null
                }
            ],
            "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/streamlit/user_info.py#L80"
        },
        "streamlit.experimental_user.values": {
            "name": "experimental_user.values",
            "signature": "st.experimental_user.values()",
            "description": "<p>D.values() -&gt; an object providing a view on D's values</p>\n",
            "args": [],
            "returns": [],
            "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/../../../../../../../Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/_collections_abc.py#L783"
        },
        "streamlit.experimental_user": {
            "name": "experimental_user",
            "signature": "st.experimental_user()",
            "is_class": true,
            "methods": [
                {
                    "name": "get",
                    "signature": "st.get.get(key, default=None)",
                    "description": "<p>D.get(k[,d]) -&gt; D[k] if k in D, else d.  d defaults to None.</p>",
                    "args": [],
                    "returns": [],
                    "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/../../../../../../../Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/_collections_abc.py#L760"
                },
                {
                    "name": "items",
                    "signature": "st.items.items()",
                    "description": "<p>D.items() -&gt; a set-like object providing a view on D's items</p>",
                    "args": [],
                    "returns": [],
                    "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/../../../../../../../Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/_collections_abc.py#L779"
                },
                {
                    "name": "keys",
                    "signature": "st.keys.keys()",
                    "description": "<p>D.keys() -&gt; a set-like object providing a view on D's keys</p>",
                    "args": [],
                    "returns": [],
                    "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/../../../../../../../Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/_collections_abc.py#L775"
                },
                {
                    "name": "to_dict",
                    "signature": "st.to_dict.to_dict()",
                    "description": "<p>Get user info as a dictionary.</p>",
                    "args": [],
                    "returns": [
                        {
                            "type_name": "Dict[str,str]",
                            "is_generator": false,
                            "description": "<p>A dictionary of the current user's information.</p>\n",
                            "return_name": null
                        }
                    ],
                    "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/streamlit/user_info.py#L80"
                },
                {
                    "name": "values",
                    "signature": "st.values.values()",
                    "description": "<p>D.values() -&gt; an object providing a view on D's values</p>",
                    "args": [],
                    "returns": [],
                    "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/../../../../../../../Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/_collections_abc.py#L783"
                }
            ],
            "properties": [],
            "source": "https://github.com/streamlit/streamlit/blob/1.30.0/lib/streamlit/user_info.py#L30",
            "description": "<p>A read-only, dict-like object for accessing information about current user.</p>\n<p><tt class=\"docutils literal\">st.experimental_user</tt> is dependant on the host platform running the\nStreamlit app. If the host platform has not configured the function, it\nwill behave as it does in a locally running app.</p>\n<p>Properties can by accessed via key or attribute notation. For example,\n<tt class=\"docutils literal\"><span class=\"pre\">st.experimental_user[&quot;email&quot;]</span></tt> or <tt class=\"docutils literal\">st.experimental_user.email</tt>.</p>\n",
            "args": [
                {
                    "name": "email",
                    "type_name": "str",
                    "is_optional": false,
                    "description": "<p>If running locally, this property returns the string literal\n<tt class=\"docutils literal\">&quot;test&#64;example.com&quot;</tt>.</p>\n<p>If running on Streamlit Community Cloud, this\nproperty returns one of two values:</p>\n<ul class=\"simple\">\n<li><tt class=\"docutils literal\">None</tt> if the user is not logged in or not a member of the app's        workspace. Such users appear under anonymous pseudonyms in the app's        analytics.</li>\n<li>The user's email if the the user is logged in and a member of the        app's workspace. Such users are identified by their email in the app's        analytics.</li>\n</ul>\n",
                    "default": null
                }
            ],
            "returns": []
        }
    }
}
```

</details>

**Comment by sfc-gh-dmatthews** [comment]
Confirmed the unwanted methods come through when using Python 3.9 but when using Python 3.11.

**Comment by sfc-gh-dmatthews** [comment]
`st.cache_data` and `st.cache_resource` now show for Streamlit in Snowflake. I have one more modification that was requested for `st.map` and `st.pydeck_chart`. I expect to get the edit request tomorrow and should be ready to merge EOD Thursday.

**Pull request merged**